### PR TITLE
Tutorial #4: The creation of Jarvis

### PR DESCRIPTION
**Pull request by Wikunia**
See #78

### CODE REVIEW

File: docs/src/tutorials/tutorial_4.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
+# **Tutorial 4:** Do you know our Mascot? - Learn about transitions and subactions!
+
+You have learned a couple of cool features of Javis already. Now you're ready to finally meet our little mascot. Well actually you can't just see him, we have to create him first. :smile:
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:3]: You'll need to manually add in the emoji. I generally use the julia repl and do `clipboard("\:smile:")` and that works.

File: docs/src/tutorials/tutorial_4.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
+# **Tutorial 4:** Do you know our Mascot? - Learn about transitions and subactions!
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:1]: Can you convert this to titlecase? Thanks!

File: docs/src/tutorials/tutorial_4.md [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+You have learned a couple of cool features of Javis already. Now you're ready to finally meet our little mascot. Well actually you can't just see him, we have to create him first. :smile:
+
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:4]: Could you add a section here about learning outcomes? Meaning, what a user will learn by following this tutorial?

File: docs/src/tutorials/tutorial_4.md [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+Let's create a list of what we want first
+- a circular head
+- some hair
+- eyes
+- a nose
+- a moving mouth
+- he should say something
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:13]: Can you make this the start of its own section?

File: docs/src/tutorials/tutorial_4.md [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+This should show you the power of `SubAction` so we want each of those appear (fade in) one after another.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md]: I'd suggest putting this line after you define the title and background so you can better and more clearly introduce a SubAction.

File: docs/src/tutorials/tutorial_4.md [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    javis(video, [
+        BackgroundAction(1:150, ground),
+        Action(title; subactions=[
+            SubAction(1:5, appear(:fade)),
+        ]),
+    ], pathname="jarvis.gif", framerate=15)
+end
+```
+
+The `ground` function should be familiar to you as well as the general structure of the `javis` function.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:46]: Put this before the code block and explain how this should be common now.

File: docs/src/tutorials/tutorial_4.md [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+
+A new part is the second `Action` which takes in a function and `subactions` as a keyword argument. 
+
+If you're wondering where the `frames` are specified:
+- You can leave the frames blank as here and then the frames from the previous action are used. 
+
+The `subactions` keyword uses a list of [`SubAction`](@ref) structs which are defined in a similar fashion with `frames` and a function but are in some sense simpler than the `Action`.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:53]: Could you explain here why we are using SubActions and why someone would want to use a SubAction?

File: docs/src/tutorials/tutorial_4.md [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+
+```julia
+Action(16:150, (args...)->circle(O, 100, :stroke); subactions=[
+    SubAction(1:15, appear(:fade)),
+]),
+```
+
+This is very similar to the previous action. Here we can see that `SubAction` uses relative frame numbers such that the head appears in the frames `16:30` and then is at full opacity afterwards.
+
+Just a small refresher: We need the anonymous function `(args...)->circle(O, 100, :stroke)` as each function gets called with the three arguments `video, action, frame`.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:73]: I'd suggest making this a Note block like this:

```mkd
> **NOTE:** blah blah
```

File: docs/src/tutorials/tutorial_4.md [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+Okay let's add some hair shall we?
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:75]: I'd suggest starting a new section here.

File: docs/src/tutorials/tutorial_4.md [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+We can use splatting for that :wink:
+
+```julia
+[
+    Action(26:150, (args...)->hair_blob(hair_angle[i]); subactions=[
+        SubAction(1:25, appear(:fade)),
+    ]) for i=1:20
+]...,
+```
+
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:108]: I really like this! :smile: However, you may want to explain what you are doing here exactly with splatting in this function.

File: docs/src/tutorials/tutorial_4.md [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+])
+```
+
+This time we also use the [`disappear`](@ref) function to fade out the text.
+
+## Final outcome and code
+
+![Jarvis](./assets/jarvis.gif)
+
+Hope you had as much fun reading this tutorial as I had creating our mascot.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:216]: Could you summarize the learning outcomes here stated at the beginning? Thanks!

File: docs/src/tutorials/tutorial_4.md [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    do_action(:stroke)
+end
+```
+
+This function uses some more functions of the awesome Luxor package.
+
+The lips should be a little thicker than the other lines that we have drawn so far so let's set `setline(2)`. (default is 1).
+First we move to the starting point of the lip and create two control points a bit below and to the vertical center.
+
+The `curve` function is used to draw a bezier curve.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:167]: I'd suggest maybe a small note on a bezier curve that links to the wikipedia article on bezier curves if anyone has questions.

File: docs/src/tutorials/tutorial_4.md (hunk 7 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md:53]: I think I explained it now in a few different ways and they are not that different from `Action`. If you think you don't understand how they work it would be great if you can clarify what you didn't understand or what users might not understand yet.

File: docs/src/tutorials/tutorial_4.md (hunk 5 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md]: I changed this to the learning outcome now so have it still at the top.

File: docs/src/tutorials/tutorial_4.md (hunk 1 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md]: thanks

File: docs/src/tutorials/tutorial_4.md (hunk 10 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md:114]: Hope it's more clear now

File: docs/src/tutorials/tutorial_4.md (hunk 12 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md]: Good point.

File: docs/src/tutorials/tutorial_4.md [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+            SubAction(1:5, appear(:fade)),
+        ]),
+    ], pathname="jarvis.gif", framerate=15)
+end
+```
+
+A new part is the second `Action` which takes in a function and `subactions` as a keyword argument this time.
+
+If you're wondering where the `frames` are specified:
+- You can leave the frames blank as here and then the frames from the previous action are used.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:51]: I would suggest this phrasing:

```mkd
A new component introduced in this tutorial is shown in the second `Action`.
It is the kwarg `subactions` followed by the object `SubAction`.
A `SubAction` is invoked when you want to apply an additional action or provide an additional functionality to the original `Action`.
Although one could use standard an `Action` to achieve the same functionality via an `id`, `subactions` provide a much easier and more organized method that also improves readability.

With the invocation of `SubAction` in the `face` function, the function can be broken down as follows:
- We define our set the background for the animation using a `BackgroundAction`.
- Using the `title` function, we invoke an `Action` and then provide a `SubAction.`
  - The `SubAction` utilizes the `Javis.jl` provided function, `appear` to cause the title to slowly appear over the `UnitRange` of `1:5`. 

> **NOTE:** For a `SubAction` you can leave the frames arg blank. 
> The frames from the previous action are used. 
```

File: docs/src/tutorials/tutorial_4.md (hunk 7 above)
Comment (TheCedarPrince) [reply docs/src/tutorials/tutorial_4.md:53]: I think this explanation is good now coupled with my suggestion in the prior conversation.

File: docs/src/tutorials/tutorial_4.md [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+- eyes
+- a nose
+- a moving mouth
+- he should say something
+
+## Learning Outcomes
+
+This tutorial should show you the power of subactions which can be used to let your objects appear and disappear as well as move on a finer grained scale.
+
+## Starting with the basics of SubAction
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:19]: Could you make each section title case?

```mkd
Starting with the basics of SubAction -> Starting with the Basics of `SubAction`
```

File: docs/src/tutorials/tutorial_4.md (hunk 10 above)
Comment (TheCedarPrince) [reply docs/src/tutorials/tutorial_4.md:114]: Perfect! Thanks Ole!

File: docs/src/tutorials/tutorial_4.md [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+    [SubAction(20i+10:20i+20, Translation(0, 5)) for i in 1:5]...
+]),
+Action(60:150, (args...)->lip(lower_lip...); subactions=[
+    SubAction(1:15, appear(:fade)),
+    [SubAction(20i:20i+10, Translation(0, 5)) for i in 1:5]...,
+    [SubAction(20i+10:20i+20, Translation(0, -5)) for i in 1:5]...
+]),
+```
+
+Yeah I like those `...` splatting :smile:
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:193]: Can you replace with the correct emoji here?

File: docs/src/tutorials/tutorial_4.md [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+
+To recap, by working through this animation you should now:
+
+1. Understand how to make objects appear and disappear using subactions
+2. Be able to move objects inside subactions to have a finer control of the movement
+3. Know how to use splatting (the `...`) to create actions in a loop
+
+Hope you had as much fun reading this tutorial as I had creating our mascot.
+
+You're now ready to create your own big project.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:233]: Just fun, maybe put a clown emoji here at the end? :clown_face:

File: docs/src/tutorials/tutorial_4.md [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+- a circular head
+- some hair
+- eyes
+- a nose
+- a moving mouth
+- he should say something
+
+## Learning Outcomes
+
+This tutorial should show you the power of subactions which can be used to let your objects appear and disappear as well as move on a finer grained scale.
```
Comment (TheCedarPrince) [inline docs/src/tutorials/tutorial_4.md:17]: I would say that a better way to lay out the learning outcomes is like this:

```mkd
This tutorial demonstrates the power of subactions.
A `SubAction` can be used to finely manipulate objects in your animation or visualization. 
From this tutorial, you will learn how to:

1. To finely control objects by making them appear and disappear using a `SubAction`.

2. Move objects using a `SubAction`.

3. Learn Julia splatting syntax (the `...`) to create multiple actions quickly.
```

File: docs/src/tutorials/tutorial_4.md (hunk 13 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md:51]: Thanks I've changed parts of it as some were not correct but added your second paragraph as a summary.

The **note** should be `Action` not `SubAction`. We currently don't provide leaving the frames out for subactions.

File: docs/src/tutorials/tutorial_4.md (hunk 16 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md:233]: I have some font issues with some of those and wasn't able to find a one which works :smile:

File: docs/src/tutorials/tutorial_4.md (hunk 15 above)
Comment (Wikunia) [reply docs/src/tutorials/tutorial_4.md:193]: Good catch

**Comment by Wikunia** [comment]
Hope this is good to go @TheCedarPrince Feel free to merge if ready!

**Comment by TheCedarPrince** [comment]
Way to go @Wikunia :smile: !
